portal-bridge/src/bridge/history.rs

```diff
@@ -277,30 +278,54 @@ impl HistoryBridge {
         sleep(Duration::from_secs(HEADER_SATURATION_DELAY)).await;
-        let timer = metrics.start_process_timer("construct_and_gossip_block_body");
-        if let Err(msg) = HistoryBridge::construct_and_gossip_block_body(
-            &full_header,
-            &portal_clients,
-            &execution_api,
-            block_stats.clone(),
-        )
-        .await
-        {
-            warn!("Error gossiping block body #{height:?}: {msg:?}");
-        };
-        metrics.stop_process_timer(timer);
-        let timer = metrics.start_process_timer("construct_and_gossip_receipts");
-        if let Err(msg) = HistoryBridge::construct_and_gossip_receipt(
-            &full_header,
-            &portal_clients,
-            &execution_api,
-            block_stats.clone(),
-        )
-        .await
-        {
-            warn!("Error gossiping block receipt #{height:?}: {msg:?}");
-        };
-        metrics.stop_process_timer(timer);
+
+        let block_body_full_header = full_header.clone();
+        let block_body_portal_clients = portal_clients.clone();
+        let block_body_execution_api = execution_api.clone();
+        let block_body_block_stats = block_stats.clone();
+        let block_body_metrics = metrics.clone();
+        let serve_block_body_task = tokio::spawn(async move {
+            match timeout(
+                SERVE_CONTENT_TIMEOUT,
+                HistoryBridge::construct_and_gossip_block_body(
+                    &block_body_full_header,
+                    &block_body_portal_clients,
+                    &block_body_execution_api,
+                    block_body_block_stats,
+                    block_body_metrics,
+                ).in_current_span(),
+            )
+            .await {
+                Ok(result) => match result {
+                    Ok(_) => (),
+                    Err(msg) => warn!("Error gossiping block body #{height:?}: {msg:?}"),
+                },
+                Err(_) => error!("construct_and_gossip_block_body() timed out on height {height}: this is an indication a bug is present")
+            };
+        });
+        let receipt_block_stats = block_stats.clone();
+        let serve_receipt_task = tokio::spawn(async move {
+            match timeout(
+                SERVE_CONTENT_TIMEOUT,
+                HistoryBridge::construct_and_gossip_receipt(
+                    &full_header,
+                    &portal_clients,
+                    &execution_api,
+                    receipt_block_stats,
+                    metrics,
+                ).in_current_span(),
+            )
+            .await {
+                Ok(result) => match result {
+                    Ok(_) => (),
+                    Err(msg) => warn!("Error gossiping block receipt #{height:?}: {msg:?}"),
+                },
+                Err(_) => error!("construct_and_gossip_receipt() timed out on height {height}: this is an indication a bug is present")
```

Comment:
Not a blocker for this pr, but this error message is very frustrating. Usually, I'll see it on my personal node after it's been running for a while. My understanding is that there's really no way to extract a relevant error message here, since the request could be blocked at any one of the many given points throughout the gossip process. But, it would be really really really cool if we could figure out some way to extract some kind of information to help understand why the timeout happened...

Reply:
Well we already know why a timeout would happen
- http client vs pandops deadlock bug
- unable to find a node to gossip 2